cmd/zrepl/zrepl.c

```diff
@@ -654,6 +755,179 @@ get_controller_ip_address(char *buf, int len)
+	hdr.opcode = ZVOL_OPCODE_HANDSHAKE;
+	hdr.len = strlen(thrd_arg->zvol_name) + 1;
+
+	rc = uzfs_zvol_socket_write(sfd, (char *)&hdr, sizeof (hdr));
+	if (rc == -1) {
+		ZREPL_ERRLOG("Socket write failed, err: %d\n", errno);
+		goto exit;
+	}
+
+	rc = uzfs_zvol_socket_write(sfd, (void *)thrd_arg->zvol_name, hdr.len);
+	if (rc == -1) {
+		ZREPL_ERRLOG("Socket write failed, err: %d\n", errno);
+		goto exit;
+	}
+
+next_step:
+	if (offset >= ZVOL_VOLUME_SIZE(zvol_state)) {
+		hdr.opcode = ZVOL_OPCODE_REBUILD_COMPLETE;
+		rc = uzfs_zvol_socket_write(sfd, (char *)&hdr, sizeof (hdr));
+		if (rc != 0) {
+			ZREPL_ERRLOG("Socket write failed, err: %d\n", errno);
+			goto exit;
+		}
+		atomic_dec_16(&zinfo->rebuild_cnt);
+		if (!zinfo->rebuild_cnt) {
+			/* Mark replica healthy now */
+			uzfs_zvol_set_rebuild_status(zinfo->zv,
+			    ZVOL_REBUILDING_DONE);
+			uzfs_zvol_set_status(zinfo->zv, ZVOL_STATUS_HEALTHY);
+		}
+		ZREPL_ERRLOG("Rebuild on LUN:%s completed\n", zinfo->name);
+		goto exit;
+	} else {
+		bzero(&hdr, sizeof (hdr));
+		hdr.status = ZVOL_OP_STATUS_OK;
+		hdr.version = REPLICA_VERSION;
+		hdr.opcode = ZVOL_OPCODE_REBUILD_STEP;
+		hdr.checkpointed_io_seq = checkpointed_io_seq;
+		hdr.offset = offset;
+		hdr.len = ZVOL_REBUILD_STEP_SIZE;
+		rc = uzfs_zvol_socket_write(sfd, (char *)&hdr, sizeof (hdr));
+		if (rc != 0) {
+			ZREPL_ERRLOG("Socket write failed, err: %d\n", errno);
+			goto exit;
+		}
+	}
+
+	while (1) {
+		rc = uzfs_zvol_socket_read(sfd, (char *)&hdr, sizeof (hdr));
+		if (rc != 0) {
+			ZREPL_ERRLOG("Socket read failed, err: %d\n", errno);
+			goto exit;
+		}
+
+		if (hdr.opcode == ZVOL_OPCODE_REBUILD_STEP_DONE) {
+			offset += ZVOL_REBUILD_STEP_SIZE;
+			printf("ZVOL_OPCODE_REBUILD_STEP_DONE received\n");
+			goto next_step;
+		}
+
+		ASSERT((hdr.opcode == ZVOL_OPCODE_READ) &&
+		    (hdr.flags & ZVOL_OP_FLAG_REBUILD));
+		hdr.opcode = ZVOL_OPCODE_WRITE;
+
+		zio_cmd = zio_cmd_alloc(&hdr, sfd);
+		rc = uzfs_zvol_socket_read(sfd, zio_cmd->buf, hdr.len);
+		if (rc != 0) {
+			zio_cmd_free(&zio_cmd);
+			ZREPL_ERRLOG("Socket read failed with "
+			    "error: %d\n", errno);
+			goto exit;
+		}
+		/* Take refcount for uzfs_zvol_worker to work on it */
+		uzfs_zinfo_take_refcnt(zinfo, B_FALSE);
+		zio_cmd->zv = zinfo;
+		taskq_dispatch(zinfo->uzfs_zvol_taskq, uzfs_zvol_worker,
+		    zio_cmd, TQ_SLEEP);
+		zio_cmd = NULL;
+	}
+
+exit:
+	kmem_free(thrd_arg, sizeof (thread_args_t));
+	if (zio_cmd != NULL)
+		zio_cmd_free(&zio_cmd);
+	if (zinfo != NULL)
+		uzfs_zinfo_drop_refcnt(zinfo, B_FALSE);
+	if (sfd != -1)
+		close(sfd);
+
+	printf("uzfs_zvol_rebuild_dw_replica thread exiting\n");
+	zk_thread_exit();
+}
+
+static int
+uzfs_zvol_rebuild_dw_replica_start(zvol_io_hdr_t *hdr, int fd, char *buf)
+{
+	int			rc, io_sfd;
+	int			rebuild_op_cnt;
+	thread_args_t		*thrd_arg;
+	mgmt_ack_t		*mgmt_ack;
+	kthread_t		*thrd_info;
+	struct sockaddr_in	replica_ip;
+
+	rebuild_op_cnt = hdr->len / sizeof (mgmt_ack_t);
```

Comment:
let us set rebuild_cnt here itself to avoid cases where rebuild of replica completes before this loop completes (due to some reason)..

Reply:
Changed name to reflect what it is.